### PR TITLE
Feature/#35 poc basic route planning with hafas

### PR DESCRIPTION
**Pull request by timvahlbrock**
### Description

Implements #35 

## Checklist

- [x] My code follows the style guidelines
- [x] I have documented my code
- [x] My changes generate no new warnings
- [x] I have linked the Issue
- [x] The Pull Request covers all the Checkpoints from the tagged Issue
- [x] I have noted the hours of work i have put in the issue

## I have edited following Parts of the project

- [x] Frontend
- [x] Backend
- [ ] Github
- [ ] Root

## If tests are required did you make them?

- [ ] No tests required
- [x] Yes
- [ ] No

### CODE REVIEW

**Comment by timvahlbrock** [comment]
I opened this pull request (in draft mode), so the progress on this is a little more visible and also there is room to discuss technical issues (e.g. the ESM module problem).

**Comment by JaJako** [comment]
Roughly speaking the backend part is finished or more specifically ready for some feedback. The frontend part of the poc will follow.

File: backend/src/hafas/hafasClient.spec.ts
Comment (JaJako) [inline backend/src/hafas/hafasClient.spec.ts:1]: Is it possible to create such a kind of playground to test what hafas-client is doing and returning? (That current version cannot be executed.)

File: backend/src/journey/journey.controller.ts
Comment (JaJako) [inline backend/src/journey/journey.controller.ts:1]: There are three endpoints (at the moment):
- /journey/          
  = plan a journey
- /journey/refresh/:token
  = refresh a journey planed before
- /journey/location/:query
  = search for a location

File: backend/src/hafas/hafasClient.ts
Comment (JaJako) [inline backend/src/hafas/hafasClient.ts:1]: There are some general HAFAS-specific settings that can be configure here (both disabled by default).
- Throttling requests (with possibility to limit request per interval)
- Retrying failed request (with customizable behaviour)

@Oeffis/core-team Should we turn on one of these?

Comment (JaJako) [inline backend/src/hafas/hafasClient.ts:1]: @Oeffis/core-team 
hafas-client's GitHub Readme states: 
> "Strictly speaking, permission is necessary to use this library with a HAFAS "mobile" endpoint. It merely tries to remove the technical barrier of accessing the data, in order to kick-start an ecosystem or apps and services that will eventually rely on [openly available data](https://opendatahandbook.org/solutions/en/Public-Transport-Data/)."

(https://github.com/public-transport/hafas-client#background)

What do we think about this? Sounds a bit like a grey area to use HAFAS without explicit permission.

File: backend/src/journey/journey.service.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+  /**
+   * Plans and returns options for the requested journey.
+   *
+   * @param journeyRequest request for a journey
+   * @return different options to plan the journey
+   */
+  planJourney(journeyRequest: JourneyRequest): Promise<Journeys> {
+    const options: JourneysOptions = {
+      results: 3,
```
Comment (JaJako) [inline backend/src/journey/journey.service.ts:38]: We have to find a good setting here. Should all results be passed to the frontend (would be "results: null") or limit the number of results?

File: backend/src/journey/journey.service.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+  /**
+   * Searches and returns available locations matching the given query.
+   *
+   * @param locationQuery query for searching locations
+   * @return locations matching the query
+   */
+  searchLocations(locationQuery: string): Promise<Locations> {
+    const options: LocationsOptions = {
+      results: 5,
```
Comment (JaJako) [inline backend/src/journey/journey.service.ts:69]: Should we limit the number of results (see above)?

File: backend/src/journey/journey.service.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  /**
+   * Plans and returns options for the requested journey.
+   *
+   * @param journeyRequest request for a journey
+   * @return different options to plan the journey
+   */
+  planJourney(journeyRequest: JourneyRequest): Promise<Journeys> {
+    const options: JourneysOptions = {
+      results: 3,
+      language: HAFAS_LANGUAGE,
```
Comment (JaJako) [inline backend/src/journey/journey.service.ts:39]: For each function of hafas-client there is a set of options. If not explicitly overwritten here defaults are used (for example: https://github.com/public-transport/hafas-client/blob/master/docs/journeys.md). 
I think the best approach would be to determine options that have to be overwritten while developing features.

File: backend/src/journey/journey.service.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    };
+
+    return this.hafas
+      .journeys(journeyRequest.from, journeyRequest.to, options);
```
Comment (JaJako) [inline backend/src/journey/journey.service.ts:43]: How to deal with errors that could get thrown by hafas-client? At the moment they are not catched in backend.

File: backend/src/hafas/hafasClient.ts
Comment (timvahlbrock) [reply backend/src/hafas/hafasClient.ts:1]: Your right. We should probably be switching to a different API. We have explicit permission to use the VRRs public EFA API, however the number of available client libraries is somewhat limited. Maybe we should talk about this in the next issue and go with the current implementation for now. I would say that we currently don't cause a serious stress on the API, but we should definitely talk about this, as we need to be able to perform a lot of requests, if we want to do statistical evaluation.

File: backend/src/journey/journey.service.ts (hunk 4 above)
Comment (timvahlbrock) [reply backend/src/journey/journey.service.ts:43]: Maybe discuss this later? So we can get this PR merged and continue with our other issues...

File: backend/src/journeys/entities/journeys.request.entity.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * Represent a request to plan a journey.
+ */
+export class JourneysRequest {
+
+  @ApiProperty({
+    description: "Where the journey starts (id of station/stop/location).",
+    required: true
+  })
+  from: StationStopLocationId;
```
Comment (JaJako) [inline backend/src/journeys/entities/journeys.request.entity.ts:15]: For now this is restricted to using an id (string).

**Review by AconexOfficial — DISMISSED** [review]

File: frontend/src/components/JourneysContainer.tsx [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  IonLabel,
+  IonList,
+  IonTextarea
+} from "@ionic/react";
+import { Journeys, Location, Station, Stop } from "hafas-client";
+import React, { useState } from "react";
+import { JourneyService, JourneysRequest } from "../api";
+
+/** Type alias for results of location query. */
+type Locations = ReadonlyArray<Station | Stop | Location>;
```
[the remark this answers is not in the document]
Comment (timvahlbrock) [reply frontend/src/components/JourneysContainer.tsx:18]: I think this is okay to the current extend, also we don't officially share types between frontend and backend.

File: backend/src/hafas/hafasClient.ts
Comment (JaJako) [reply backend/src/hafas/hafasClient.ts:1]: see #154

File: backend/src/journey/journey.service.ts (hunk 4 above)
Comment (JaJako) [reply backend/src/journey/journey.service.ts]: see #156

File: backend/src/journeys/controller/journeysController.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      .planJourney(journeyRequest);
+  }
+
+  @Get("refresh/:token")
+  @ApiOperation({
+    summary: "Returns refreshed data of a journey.",
+    description: "Refresh (receive updated data of) a journey that has been planned before."
+  })
+  @ApiOkResponse({
+    description: "Returns refreshed journey.",
```
[the remark this answers is not in the document]
Comment (JaJako) [reply backend/src/journeys/controller/journeysController.ts:38]: see #157

File: backend/src/journeys/entities/journeys.request.entity.ts (hunk 5 above)
Comment (JaJako) [reply backend/src/journeys/entities/journeys.request.entity.ts:15]: see #158

**Comment by JaJako** [comment]
Rebased (some adjustments were necessary because following commit 3f0376fd).

File: frontend/src/components/JourneysContainer.tsx [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+  IonCardHeader,
+  IonCardSubtitle,
+  IonCardTitle,
+  IonInput,
+  IonItem,
+  IonLabel,
+  IonList,
+  IonTextarea
+} from "@ionic/react";
+import { Journeys, Location, Station, Stop } from "hafas-client";
```
Comment (timvahlbrock) [inline frontend/src/components/JourneysContainer.tsx:13]: The usage of types from a backend dependency is due to the problem with the return type definitions you mentioned earlier, right? Can you add to #157 that when it's done this dependency should also be removed from the frontend?

**Comment by timvahlbrock** [comment]
@JaJako This looks good to me, but I can't approve this PR myself, because I opened it. Kinda messed that up, sorry.

File: frontend/src/components/JourneysContainer.tsx (hunk 8 above)
Comment (JaJako) [reply frontend/src/components/JourneysContainer.tsx:13]: I haven´t had that in mind, but you´re right: Once the definition of API return types is working correctly the '@types/hafas-client' dependency in frontend can (and should) be removed.

**Comment by timvahlbrock** [comment]
@JaJako I think you can approve them yourself too. I took a look at the changed files so two sets of eyes concept is satisfied

**Review by JaJako — APPROVED** [review]

**Pull request merged**